File: CoreJava/algorithms/src/main/java/org/gb/sample/algo/intro_to_algo_book/matrix_multiplication/SerialBottomUpMatrixChainMultiplier.java
package org.gb.sample.algo.intro_to_algo_book.matrix_multiplication;

import java.util.List;

import org.javatuples.Pair;

class SerialBottomUpMatrixChainMultiplier extends AbstractMatrixChainMultiplier {

	private void initOptimalProductArr(final List<Matrix> matrixChain, final int matrixCount, final Matrix[][] optimalProductArr) {
		for (int chainIdx = 0; chainIdx < matrixCount; chainIdx++) {
			optimalProductArr[chainIdx][chainIdx] = matrixChain.get(chainIdx);
		}
	}

	private Pair<Integer, Integer> computeOptimalSplit(final List<Matrix> matrixChain, final int sgmtBeginIdx, final int sgmtEndIdx,
			final int[][] minMultiplyCountArr, int minMultiplyCount) {
		int optimalSplitIdx = sgmtBeginIdx;
		for (int splitIdx = sgmtBeginIdx; splitIdx < sgmtEndIdx; splitIdx++) {
			Matrix sgmtHead = matrixChain.get(sgmtBeginIdx);
			Matrix sgmtSplit = matrixChain.get(splitIdx);
			Matrix sgmtTail = matrixChain.get(sgmtEndIdx);
			int subsgmtsMultiplyCount = sgmtHead.getRowCount() * sgmtSplit.getColumnCount() * sgmtTail.getColumnCount();
			int cumltvMultiplyCount = minMultiplyCountArr[sgmtBeginIdx][splitIdx] + minMultiplyCountArr[splitIdx + 1][sgmtEndIdx] + subsgmtsMultiplyCount;

			if (cumltvMultiplyCount < minMultiplyCount) {
				optimalSplitIdx = splitIdx;
				minMultiplyCount = cumltvMultiplyCount;
			}
		}
		return Pair.with(optimalSplitIdx, minMultiplyCount);
	}

	@Override
	Matrix computeOptimalOrderInternal(final List<Matrix> matrixChain) {
		final int matrixCount = matrixChain.size();
		final Matrix[][] optimalProductArr = new Matrix[matrixCount][matrixCount];
		initOptimalProductArr(matrixChain, matrixCount, optimalProductArr);

		final int[][] minMultiplyCountArr = new int[matrixCount][matrixCount];
		for (int segmentLength = 2; segmentLength <= matrixCount; segmentLength++) {
			for (int sgmtBeginIdx = 0; sgmtBeginIdx <= (matrixCount - segmentLength); sgmtBeginIdx++) {
				int sgmtEndIdx = sgmtBeginIdx + segmentLength - 1;
				Pair<Integer, Integer> optimalSplitResult = computeOptimalSplit(matrixChain, sgmtBeginIdx, sgmtEndIdx, minMultiplyCountArr, Integer.MAX_VALUE);

				int optimalSplitIdx = optimalSplitResult.getValue0().intValue();
				int minMultiplyCount = optimalSplitResult.getValue1().intValue();
				Matrix sgmtHead = matrixChain.get(sgmtBeginIdx);
				Matrix sgmtTail = matrixChain.get(sgmtEndIdx);
				Matrix leftSubsgmtOptimalProduct = optimalProductArr[sgmtBeginIdx][optimalSplitIdx];
				Matrix rightSubsgmtOptimalProduct = optimalProductArr[optimalSplitIdx + 1][sgmtEndIdx];

				minMultiplyCountArr[sgmtBeginIdx][sgmtEndIdx] = minMultiplyCount;
				Matrix sgmtOptimalProduct = new Matrix(leftSubsgmtOptimalProduct, rightSubsgmtOptimalProduct, sgmtHead.getRowCount(), sgmtTail.getColumnCount(),
						minMultiplyCount);
				optimalProductArr[sgmtBeginIdx][sgmtEndIdx] = sgmtOptimalProduct;
			}
		}

		return optimalProductArr[0][matrixCount - 1];
	}

}

File: CoreScala/simple-scala/src/main/scala/org/gb/sample/scala/intro_to_algo_book/matrix_multiplication/SerialBottomUpMatrixChainMultiplier.scala
package org.gb.sample.scala.intro_to_algo_book.matrix_multiplication

class SerialBottomUpMatrixChainMultiplier extends AbstractMatrixChainMultiplier {

    private def initOptimalProductArr(matrixChain: Seq[Matrix], matrixCount: Int, optimalProductArr: Array[Array[Matrix]]): Unit = {
        for (chainIdx <- 0 until matrixCount) {
            optimalProductArr(chainIdx)(chainIdx) = matrixChain(chainIdx)
        }
    }

    private def computeOptimalSplit(matrixChain: Seq[Matrix], sgmtBeginIdx: Int, sgmtEndIdx: Int, minMultiplyCountArr: Array[Array[Int]]): (Int, Int) = {
        var optimalSplitIdx = sgmtBeginIdx
        var minMultiplyCount = Int.MaxValue
        for (splitIdx <- sgmtBeginIdx until sgmtEndIdx) {
            val sgmtHead = matrixChain(sgmtBeginIdx)
            val sgmtSplit = matrixChain(splitIdx)
            val sgmtTail = matrixChain(sgmtEndIdx)
            val subsgmtsMultiplyCount = sgmtHead.rowCount * sgmtSplit.columnCount * sgmtTail.columnCount
            val cumltvMultiplyCount = minMultiplyCountArr(sgmtBeginIdx)(splitIdx) + minMultiplyCountArr(splitIdx + 1)(sgmtEndIdx) + subsgmtsMultiplyCount
            if(cumltvMultiplyCount < minMultiplyCount) {
                optimalSplitIdx = splitIdx
                minMultiplyCount = cumltvMultiplyCount
            }
        }
        return (optimalSplitIdx, minMultiplyCount)
    }

    override def computeOptimalOrderInternal(matrixChain: Seq[Matrix]): Matrix = {
        val matrixCount = matrixChain.size
        val optimalProductArr = Array.ofDim[Matrix](matrixCount, matrixCount)
        initOptimalProductArr(matrixChain, matrixCount, optimalProductArr)

        val minMultiplyCountArr = Array.ofDim[Int](matrixCount, matrixCount)
        for (segmentLength <- 2 to (matrixCount)) {
            for (sgmtBeginIdx <- 0 to (matrixCount - segmentLength)) {
                val sgmtEndIdx = sgmtBeginIdx + segmentLength - 1
                val (optimalSplitIdx, minMultiplyCount) = computeOptimalSplit(matrixChain, sgmtBeginIdx, sgmtEndIdx, minMultiplyCountArr)
                val sgmtHead = matrixChain(sgmtBeginIdx)
                val sgmtTail = matrixChain(sgmtEndIdx)
                val leftSubsgmtOptimalProduct = optimalProductArr(sgmtBeginIdx)(optimalSplitIdx)
                val rightSubsgmtOptimalProduct = optimalProductArr(optimalSplitIdx + 1)(sgmtEndIdx)

                minMultiplyCountArr(sgmtBeginIdx)(sgmtEndIdx) = minMultiplyCount
                val sgmtOptimalProduct = Matrix(leftSubsgmtOptimalProduct, rightSubsgmtOptimalProduct, sgmtHead.rowCount,
                    sgmtTail.columnCount, minMultiplyCount)
                optimalProductArr(sgmtBeginIdx)(sgmtEndIdx) = sgmtOptimalProduct
            }
        }

        return optimalProductArr(0)(matrixCount - 1)
    }
}
